snuba/redis.py

```diff
@@ -72,29 +77,80 @@ def wrapper() -> RedisClientType:
     return _retry_inner
 
 
+_EXISTING_CLIENTS_BY_FUNCTION: MutableMapping[ClusterFunction, RedisClientType] = {}
+_EXISTING_CLIENTS_BY_CONNECTION_PARAMS: MutableMapping[
+    ConnectionDescriptor, RedisClientType
+] = {}
+
+_cluster_list_initialized = False
+
+
+@_retry(max_retries=settings.REDIS_INIT_MAX_RETRIES)
+def _build_single_node_client(
+    connection_descriptor: ConnectionDescriptor,
+) -> RedisClientType:
+    assert isinstance(connection_descriptor, NodeDescriptor)
+    return StrictRedis(
+        host=connection_descriptor.host,
+        port=connection_descriptor.port,
+        password=connection_descriptor.password,
+        db=connection_descriptor.db,
+        socket_keepalive=True,
+    )
+
+
 @_retry(max_retries=settings.REDIS_INIT_MAX_RETRIES)
-def _initialize_redis_cluster() -> RedisClientType:
-    if settings.USE_REDIS_CLUSTER:
-        startup_nodes = settings.REDIS_CLUSTER_STARTUP_NODES
-        if startup_nodes is None:
-            startup_nodes = [{"host": settings.REDIS_HOST, "port": settings.REDIS_PORT}]
-        startup_cluster_nodes = [
-            ClusterNode(n["host"], n["port"]) for n in startup_nodes
-        ]
-        return RetryingStrictRedisCluster(
-            startup_nodes=startup_cluster_nodes,
-            socket_keepalive=True,
-            password=settings.REDIS_PASSWORD,
-            max_connections_per_node=True,
-        )
-    else:
-        return StrictRedis(
-            host=settings.REDIS_HOST,
-            port=settings.REDIS_PORT,
-            password=settings.REDIS_PASSWORD,
-            db=settings.REDIS_DB,
-            socket_keepalive=True,
-        )
-
-
-redis_client: RedisClientType = _initialize_redis_cluster()
+def _build_cluster_client(
+    connection_descriptor: ConnectionDescriptor,
+) -> RedisClientType:
+    assert isinstance(connection_descriptor, Sequence)
```

Comment:
If we want to be this strict about the typing, then `ConnectionDescriptor` should be a class that validates its own input.